fabric-network/test/impl/event/blocklistener.ts

```diff
@@ -0,0 +1,214 @@
+		eventService.sendEvent(newEvent(1));
+
+		await listener.completePromise;
+		sinon.assert.calledOnce(spy);
+	});
+
+	it('remove listener multiple times has no effect', async () => {
+		const removedListener = sinon.spy(testUtils.newAsyncListener<BlockEvent>());
+
+		await network.addBlockListener(listener);
+		await network.addBlockListener(removedListener);
+		network.removeBlockListener(removedListener);
+		network.removeBlockListener(removedListener);
+		eventService.sendEvent(newEvent(1));
+
+		await listener.completePromise;
+		sinon.assert.notCalled(removedListener);
+	});
+
+	it('listener can remove itself when receiving event', async () => {
+		listener = testUtils.newAsyncListener<BlockEvent>(2);
+		const fake = sinon.fake(async (event: BlockEvent) => {
+			network.removeBlockListener(fake);
+		});
+
+		await network.addBlockListener(listener);
+		await network.addBlockListener(fake);
+		eventService.sendEvent(newEvent(1));
+		// fake listener should have removed itself
+		eventService.sendEvent(newEvent(2));
+
+		await listener.completePromise;
+		sinon.assert.calledOnce(fake);
+	});
+
+	it('listener does not auto-unregister when receiving events', async () => {
+		listener = testUtils.newAsyncListener<BlockEvent>(2);
+		const event1 = newEvent(1);
+		const event2 = newEvent(2);
+
+		await network.addBlockListener(listener);
+		eventService.sendEvent(event1);
+		eventService.sendEvent(event2);
+
+		const actual = await listener.completePromise;
+		network.removeBlockListener(listener);
+
+		expect(actual).to.deep.equal([event1, event2]);
+	});
+
+	it('error thrown by listener is handled', async () => {
+		listener = testUtils.newAsyncListener<BlockEvent>(2);
+		const errorListener = sinon.fake.rejects(new Error('LISTENER_ERROR'));
+		const event1 = newEvent(1);
+		const event2 = newEvent(2);
+
+		await network.addBlockListener(listener);
+		await network.addBlockListener(errorListener);
+		eventService.sendEvent(event1);
+		eventService.sendEvent(event2);
+
+		const actual = await listener.completePromise;
+		expect(actual).to.deep.equal([event1, event2]);
+	});
+
+	it('listener receives blocks in order', async () => {
+		listener = testUtils.newAsyncListener<BlockEvent>(3);
+		const event1 = newEvent(1);
+		const event2 = newEvent(2);
+		const event3 = newEvent(3);
+
+		await network.addBlockListener(listener);
+		eventService.sendEvent(event1);
+		eventService.sendEvent(event3);
+		eventService.sendEvent(event2);
+
+		const actual = await listener.completePromise;
+		expect(actual).to.deep.equal([event1, event2, event3]);
+	});
+
+	it('listener does not receive old blocks', async () => {
+		listener = testUtils.newAsyncListener<BlockEvent>(2);
+		const event1 = newEvent(1);
+		const event2 = newEvent(2);
+		const event3 = newEvent(3);
+
+		await network.addBlockListener(listener);
+		eventService.sendEvent(event2);
+		eventService.sendEvent(event1); // Ignored as older than first block received
```

Comment:
We will have to warn the user that if the first block is received out of order they will miss blocks

Reply:
This is only for realtime listening. In that case you get all blocks (in order) with a larger block number than the first one the client sees. For replay you get from exactly the block you specified. Nothing gets missed.